vclock capability

chardan opened the pull request:
Adds a new capability to support different encoding options for vector clocks. Moves existing zlib support to a capability, implements raw encoding. The purpose is to allow users to better tune for compression or performance.

Commit:
Support for vector clock encoding capability.

Commit:
Cleaning up, corrected call names.

Commit:
Missed a few encoding locations; trying to move them all into one place.

Commit:
Tracking down encoding/decoding bug.

Commit:
Fix vclock type error.

Commit:
Move kv_wm utils changes to riak_object; make dialyzer a bit more happy.

chardan commented:
See-also accompanying tests in riak_test br jfw-vclock_capability:
https://github.com/basho/riak_test/pull/233

engelsanchez commented on src/riak_object.erl:
```diff
@@ -389,10 +390,17 @@ assemble_index_specs(Indexes, IndexOp) ->
+%% @spec vclock_header(riak_object()) -> {Name::string(), Value::string()}
+%% @doc Transform the Erlang representation of the document's vclock
+%%      into something suitable for an HTTP header
+vclock_header(Doc) ->
```
Moving this method here doesn't seem that clean to me.  We now have a bit of web machine http knowledge embedded in the internal riak object representation.  I think that goes beyond the scope of this fix.

engelsanchez commented on src/riak_object.erl:
```diff
@@ -541,6 +549,31 @@ syntactic_merge(CurrentObject, NewObject) ->
+    riak_core_capability:get({riak_kv, vclock_data_encoding}, encode_zlib).
+
+%% Encode a vclock in accordance with our capability setting:
+-spec encode_vclock(VClock :: term()) -> binary().
```
So, it seems from above that this is really is taking a binary, not a term. Also, a plain string will work just due to the fact that zlib:zip and base64:encode both would take it (not that we need to advertise this internal detail in the spec).

engelsanchez commented on src/riak_object.erl:
```diff
@@ -541,6 +549,31 @@ syntactic_merge(CurrentObject, NewObject) ->
+    end.
+
+%% Decode a vclock against our capability settings:
+-spec decode_vclock(VClock :: term()) -> vclock:vclock().
```
Same here for the input as above (term() does not seem right), and we are returning a binary()

engelsanchez commented on src/riak_object.erl:
```diff
@@ -389,10 +390,17 @@ assemble_index_specs(Indexes, IndexOp) ->
 %% @spec to_json(riak_object()) -> {struct, list(any())}
 %% @doc Converts a riak_object into its JSON equivalent
 to_json(Obj=#r_object{}) ->
-    {_,Vclock} = riak_kv_wm_utils:vclock_header(Obj),
```
I see now why the vclock_header method was moved here.  Still, we have the typical dilemma of the simplicity of having all of the encoding stuff built into the basic object and the cleanliness and flexibility of having anything to do with encoding moved to a different module.

engelsanchez commented on src/riak_kv_wm_object.erl:
```diff
@@ -815,18 +815,16 @@ encode_vclock_header(RD, #ctx{doc={ok, Doc}}) ->
-
-encode_vclock(VClock) ->
-    binary_to_list(base64:encode(zlib:zip(term_to_binary(VClock)))).
+    %% JFW: I'm slightly suspicious about this because we're dropping the initial term_to_binary(Doc):
```
The original vclock_header took a riak_object and produced the header tuple. This can't be right. Now Doc is expected to be a binary version of the vclock. Unless something else has changed upstream, this is definitely different.

engelsanchez commented:
Also, I would recommend merging from master. I'm doing that now locally to be able to test this. There are some conflicts.

Commit:
More bugfixing-- fixes "sis/bro" test, working on Java client regression.

Commit:
Fix list:keytake() error in Java client unit tests; correct encoding
dyalyzer spec.

Commit:
Remove comments.

engelsanchez commented:
@chardan were my comments on this PR addressed with the last changes? Can I continue reviewing this?  Don't forget to comment on a PR if you do address review comments so the reviewer gets notified and knows when to continue.

Commit:
Remove comment and add better dialyzer spec.

Commit:
Remove comments.

chardan commented:
I believe so, have a look!

engelsanchez commented on src/riak_kv_pb_object.erl:
```diff
@@ -293,11 +293,11 @@ erlify_rpbvc(undefined) ->
     vclock:fresh();
 erlify_rpbvc(PbVc) ->
-    binary_to_term(zlib:unzip(PbVc)).
+    binary_to_term(riak_object:decode_vclock(PbVc)).
```
Because of this change, the representation of a vclock through the PB interface changes from a binary to a base64 encoded binary (see riak_object:decode_vclock).  So, if a node is upgraded in a cluster, vclocks fetched from other nodes will be incompatible with this node. For instances where the vclock is sent to a different node than the original one, I believe this will fail.  This is in addition to requests failing when getting a vclock from a cluster right before the capabilities change to a new format, which we should probably handle too. Thoughts?

chardan replied:
Very good observation, I'll address it. Thanks!

engelsanchez replied:
Unfortunately the way this was addressed does not fix the problem. See this commit for an updated riak_test that shows how it fails in a mixed cluster when going from a new node to an old node.  Ideally, vclocks through PB should be encoded exactly the same way in a mixed cluster than they were in the old cluster, even though that makes things more complicated because PB was encoding in a different way to the HTTP interface, which messes with the refactoring to clean it up.

Commit:
Fallback and encoding method support for vector clocks.

Commit:
Great Merge Disaster fixes; restored exception handling.

Commit:
More reconcilation for rebase.

Commit:
debugging

Commit:
Fix mixed current/legacy clocks.

engelsanchez commented on src/riak_object.erl:
```diff
@@ -937,7 +1001,6 @@ new_with_ctype_test() ->
     O = riak_object:new(<<"b">>, <<"k">>, <<"{\"a\":1}">>, "application/json"),
     ?assertEqual("application/json", dict:fetch(?MD_CTYPE, riak_object:get_metadata(O))).
 
-new_with_md_test() ->
```
@chardan, this is not compiling. I see the aftermath of a rebase here. There is no merge history to figure out what else broke with the rebase.  Please correct and re-check if you have lost anything else if you can.  Please be careful not to check in broken code man. Compile and run the tests after a merge/rebase.

chardan replied:
Apologies for the rebase. I missed the conditional compilation, it's been corrected, along with bugfixes for the ancestor/sibling test.

Commit:
Fix b0rked compile during eunit; fix bro/sis test (again)

Commit:
Fix failure on put() to multiple same-versioned clients; fix Java client
failure.

engelsanchez committed:
Fix mixed cluster vclock encoding problems

This fix does a few things:
* Removes base64 from the riak_object codec functions. It's used by code
  that wants to send the binary encoded vclock over text explicitly
(JSON and HTTP).
* Normalizes interface to codecs: you encode and actual vclock and get a
  binary, and decoding this binary returns a vclock, however they are
  implemented.
* Added vclock codec round trip unit tests

engelsanchez commented:
I have attempted to fix all problems I reported with this commit https://github.com/basho/riak_kv/commit/6a05f3706e6be87f46d55ac5526567e2b2f7baa1

It also takes care of making sure we only binary to term/term to binary raw encoded values once.  Jesse will be on vacation for a bit, so we'll need someone else to review my fixes.
